lib/stoplight/light/runnable.rb

```diff
@@ -24,63 +24,71 @@ def color
         end
       end
 
+      # @param fallback [#call, nil] (nil) a fallback block to be called when the light is red
       # @raise [Error::RedLight]
-      def run(&code)
-        raise ArgumentError, <<~ERROR unless block_given?
-          nothing to run. Please, pass a block into `Light#run`
-        ERROR
+      def run(fallback = nil, &code)
+        raise ArgumentError, 'nothing to run. Please, pass a block into `Light#run`' unless block_given?
 
         case color
-        when Color::GREEN then run_green(&code)
-        when Color::YELLOW then run_yellow(&code)
-        else run_red
+        when Color::GREEN then run_green(fallback, &code)
+        when Color::YELLOW then run_yellow(fallback, &code)
+        else run_red(fallback)
         end
       end
 
       private
 
-      def run_green(&code)
+      def run_green(fallback, &code)
         on_failure = lambda do |size, error|
           notify(Color::GREEN, Color::RED, error) if failures_threshold_breached?(size, threshold)
         end
-        run_code(nil, on_failure, &code)
+        run_code(nil, on_failure, fallback, &code)
       end
 
       def failures_threshold_breached?(current_failures_count, max_errors_threshold)
         current_failures_count == max_errors_threshold
       end
 
-      def run_yellow(&code)
+      def run_yellow(fallback, &code)
         on_success = lambda do |failures|
           notify(Color::RED, Color::GREEN) unless failures.empty?
         end
-        run_code(on_success, nil, &code)
+        run_code(on_success, nil, fallback, &code)
       end
 
-      def run_red
+      # @param fallback [#call, nil]
+      def run_red(fallback)
         raise Error::RedLight, name unless fallback
 
         fallback.call(nil)
       end
 
-      def run_code(on_success, on_failure, &code)
+      def run_code(on_success, on_failure, fallback, &code)
         result = code.call
         failures = clear_failures
```

Comment:
It looks like we do not check for the time window if we use in-memory storage, so there might be a bug here, but I'm not sure about this. I'll take a closer look at it tomorrow 🙂

Reply:
Thanks a bunch 🙇